test/extensions/filters/network/local_ratelimit/local_ratelimit_integration_test.cc

```diff
@@ -0,0 +1,61 @@
+#include "test/integration/integration.h"
+
+namespace Envoy {
+namespace {
+
+class LocalRateLimitIntegrationTest : public Event::TestUsingSimulatedTime,
+                                      public testing::TestWithParam<Network::Address::IpVersion>,
+                                      public BaseIntegrationTest {
+public:
+  LocalRateLimitIntegrationTest()
+      : BaseIntegrationTest(GetParam(), ConfigHelper::TCP_PROXY_CONFIG) {}
+
+  ~LocalRateLimitIntegrationTest() override {
+    test_server_.reset();
+    fake_upstreams_.clear();
+  }
+
+  void setup(const std::string& filter_yaml) {
+    config_helper_.addNetworkFilter(filter_yaml);
+    BaseIntegrationTest::initialize();
+  }
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, LocalRateLimitIntegrationTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+// Make sure the filter works in the basic case.
+TEST_P(LocalRateLimitIntegrationTest, NoRateLimiting) {
+  setup(R"EOF(
+name: envoy.filters.network.local_ratelimit
+typed_config:
+  "@type": type.googleapis.com/envoy.config.filter.network.local_rate_limit.v2alpha.LocalRateLimit
+  stat_prefix: local_rate_limit_stats
+  token_bucket:
+    max_tokens: 1
+    fill_interval: 0.2s
+)EOF");
+
+  IntegrationTcpClientPtr tcp_client = makeTcpConnection(lookupPort("listener_0"));
+  FakeRawConnectionPtr fake_upstream_connection;
+  ASSERT_TRUE(fake_upstreams_[0]->waitForRawConnection(fake_upstream_connection));
+  tcp_client->write("hello");
+  ASSERT_TRUE(fake_upstream_connection->waitForData(5));
+  ASSERT_TRUE(fake_upstream_connection->write("world"));
+  tcp_client->waitForData("world");
+  tcp_client->close();
+  ASSERT_TRUE(fake_upstream_connection->waitForDisconnect(true));
+
+  EXPECT_EQ(0,
+            test_server_->counter("local_rate_limit.local_rate_limit_stats.rate_limit")->value());
+}
+
+// TODO(mattklein123): Create an integration test that tests rate limiting. Right now this is
+// not easily possible using simulated time due to the fact that simulated time runs alarms on
+// their correct threads when woken up, but does not have any barrier for when the alarms have
+// actually fired. This makes a deterministic test impossible without resorting to hacks like
+// storing the number of tokens in a stat, etc.
```

Comment:
@jmarantz this is the test that I want to add using simulated time, but there is no way to do it per the comment. This would be a great specific case to use to add the barrier behavior?

Reply:
Do you want all threads blocked? Or just the main thread? Could we use a condvar or absl::Notification, which blocks until a timer on a worker-thread signals it?

Is this something that needs to be in the simulated time abstraction? Or could it be layered on? Are the timers you want to block on set by production code? Or are they test-specific?

Reply:
The timer I want to wait on is the token bucket refill timer, which is in prod code running on the main thread. I think the way I would like this to work optimally is:
1) When we advance sim time we keep track of all timers we end up waking up.
2) Have some barrier/condvar that is notified when the timers actually wake up and run their callback.
3) Block the caller of simtime advance until all timers have run.

Obviously we would need to account for timers that run on the test thread, but this doesn't seem that bad.